torch/csrc/api/include/torch/nn/parallel/data_parallel.h

```diff
@@ -0,0 +1,161 @@
+    // in `torch::ones({3, 4})`, the device is unspecified and pulled from the
+    // current thread local default options. As such, we can here modify these
+    // thread local default options and thereby cause all tensors in the cloned
+    // module to be constructed directly on the device we want.
+    OptionsGuard guard(device);
+    replicas.push_back(std::static_pointer_cast<ModuleType>(module->clone()));
+  }
+  return replicas;
+}
+
+/// Replicates a module holder on the given list of devices.
+/// This method allows calling `replicate()` with a module holder, such as
+/// `Linear`.
+template <typename ModuleType>
+std::vector<ModuleHolder<ModuleType>> replicate(
+    const ModuleHolder<ModuleType>& module,
+    const std::vector<Device>& devices) {
+  auto ptrs = replicate(module.ptr(), devices);
+  return std::vector<ModuleHolder<ModuleType>>(ptrs.begin(), ptrs.end());
+}
+
+/// Applies the given inputs to the given modules in a parallel fashion.
+/// Conceptually, a thread is spawned for each `(module, input)` pair, in which
+/// `forward()` is called on the module with its corresponding input. The
+/// outputs of the individual calls are stored in a vector and returned.
+///
+/// Further remarks:
+/// 1. The length of the module container must match the length of the inputs.
+/// 2. If a list of devices is supplied, it must match the list of modules in
+/// length. Each device will be set to the current default device during the
+/// invocation of the respective module. This means any tensors allocated on the
+/// default device inside the module will be constructed on this device.
+template <typename ModuleType>
+std::vector<Tensor> parallel_apply(
+    std::vector<ModuleType>& modules,
+    const std::vector<Tensor>& inputs,
+    const at::optional<std::vector<Device>>& devices = at::nullopt) {
+  AT_CHECK(
+      modules.size() == inputs.size(), "Must have as many inputs as modules");
+  if (devices) {
+    AT_CHECK(
+        modules.size() == devices->size(),
+        "Must have as many devices as modules");
+  }
+
+  std::vector<Tensor> outputs(modules.size());
+  std::mutex mutex;
+
+  at::parallel_for(
+      /*begin=*/0,
+      /*end=*/modules.size(),
+      /*grain_size=*/1,
+      [&modules, &inputs, &devices, &outputs, &mutex](
+          int64_t index, int64_t stop) {
+        for (; index < stop; ++index) {
+          try {
+            torch::OptionsGuard options_guard(
+                devices ? (*devices)[index] : inputs[index].device());
+            auto output = modules[index]->forward(inputs[index]);
+            std::lock_guard<std::mutex> lock(mutex);
+            outputs[index] = output;
+          } catch (const std::exception& exception) {
+            std::cerr << "Replica #" << index
+                      << " terminated in parallel_apply() with exception: "
+                      << exception.what();
+            std::terminate();
```

Comment:
That's not too elegant (both printing and terminating unconditionally). We stash the exception in Python land and reraise it later. I think it's possible even in C++ using `std::exception_ptr`